src/irmin-pack/reconstruct_index.ml

```diff
@@ -16,98 +47,121 @@ module Make (Args : Args) : sig
+      let off_after_key, key = decode_key buffer buffer_off in
+      assert (off_after_key = buffer_off + Hash.hash_size);
+      let off_after_kind, kind = decode_kind buffer off_after_key in
+      assert (off_after_kind = buffer_off + Hash.hash_size + 1);
+      (* Get the length of the entire entry *)
+      let entry_len = decode_entry_length kind buffer buffer_off in
+      { key; data = (off, entry_len, kind) }
+    with
+    | Invalid_argument msg when msg = "index out of bounds" ->
+        raise Not_enough_buffer
+    | Invalid_argument msg when msg = "String.blit / Bytes.blit_string" ->
+        raise Not_enough_buffer
+
+  let ingest_data_file ~progress ~total pack index =
+    let buffer = ref (Bytes.create 1024) in
+    let refill_buffer ~from =
+      let read = IO.read pack ~off:from !buffer in
+      assert (read = Bytes.length !buffer)
     in
-    let rec read_and_decode ?(retries = 1) off =
-      Log.debug (fun l ->
-          l "read_and_decode retries %d off %a" retries Int63.pp off);
-      let chunk = 64 * 10 * retries in
-      let buf = IO.read_buffer ~chunk ~off pack in
-      match decode_entry buf off with
-      | Some new_off -> new_off
-      | None ->
-          (* the biggest entry in a tezos store is a blob of 54801B *)
-          if retries > 90 then
-            failwith "too many retries to read data, buffer size = 57600B"
-          else (read_and_decode [@tailcall]) ~retries:(retries + 1) off
+    let expand_and_refill_buffer ~from =
+      let length = Bytes.length !buffer in
+      if length > 1_000_000_000 (* 1 GB *) then
+        failwith
+          "Couldn't decode a value in %d of buffer space. Corrupted data file?"
+      else (
+        buffer := Bytes.create (2 * length);
+        refill_buffer ~from)
     in
-    let rec read_buffer off =
-      if off >= total then ()
+    let stats = Stats.empty () in
+    let rec loop_entries ~buffer_off off =
+      if off >= total then stats
       else
-        let new_off = read_and_decode off in
-        (read_buffer [@tailcall]) new_off
+        let buffer_off, off =
+          match
+            decode_entry_exn ~off
+              ~buffer:(Bytes.unsafe_to_string !buffer)
+              ~buffer_off
+          with
+          | { key; data } ->
+              let off', entry_len, kind = data in
+              let entry_lenL = Int63.of_int entry_len in
+              assert (off = off');
+              Log.debug (fun l ->
+                  l "k = %a (off, len, kind) = (%a, %d, %a)" pp_key key Int63.pp
+                    off entry_len Pack_value.Kind.pp kind);
+              Stats.add stats kind;
+              Index.add index key data;
+              progress entry_lenL;
+              (buffer_off + entry_len, off ++ entry_lenL)
+          | exception Not_enough_buffer ->
+              let () =
+                if buffer_off > 0 then
+                  (* Try again with the value at the start of the buffer. *)
```

Comment:
I don't understand when this case occurs; we did not have it in the previous index reconstruction, no?

Reply:
We did not indeed. The previous reconstruction used a fresh buffer for every attempted decode, so was always hitting the case `buffer_off = 0` here. In practice, this is a bit wasteful because (a) we can keep the same buffer throughout, and (b) it's possible that many entries will fit contiguously in that buffer (and we can get them all with one file read). This leads to a case where we've exhausted the remaining buffer space but the value we're trying to read might still fit inside the entire buffer:

![image](https://user-images.githubusercontent.com/26125912/121889708-7bcb6f00-cd11-11eb-8c93-25409dd0230f.png)

(In practice, this is hopefully the common case.)

Reply:
I see, thanks for the explanation and the drawing :)